src/Engine/EmitMSIL/BuiltIn.cs

```diff
@@ -0,0 +1,383 @@
+﻿using ProtoCore.DSASM;
+using ProtoCore.Utils;
+using System;
+using System.Reflection;
+
+namespace EmitMSIL
+{
+    public class BuiltIn
+    {
+        private static object ToBoolean(object a)
+        {
+            switch (a)
+            {
+                case bool ab:
+                    return ab;
+
+                case long al:
+                    return al != 0;
+
+                case int ai:
+                    return ai != 0;
+
+                case null:
+                    return false;
+
+                case double ad:
+                    return !Double.IsNaN(ad) && !ad.Equals(0.0);
+
+                case string str:
+                    return !string.IsNullOrEmpty(str);
+
+                case char ac:
+                    return ac != 0;
+            }
+
+            if (a is Pointer)
+            {
+                return true;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Handles possible overflows from a checked operation. If it works, its result is
+        /// returned, otherwise the result of the unchecked operation is returned and a warning
+        /// is logged.
+        /// </summary>
+        /// <param name="checkedOperation">Checked operation to be attempted first</param>
+        /// <param name="uncheckedOperation">Unchecked operation to be perfomed when the checked operation overflowed</param>
+        /// <returns>The result of the first succesful operation</returns>
+        private static T HandleOverflow<T>(Func<T> checkedOperation, Func<T> uncheckedOperation)
+        {
+            try
+            {
+                return checkedOperation();
+            }
+            catch (OverflowException)
+            {
+//                runtimeCore.RuntimeStatus.LogWarning(WarningID.IntegerOverflow, string.Format($"{Resources.IntegerOverflow}href=IntegerOverflow.html"));
+                return uncheckedOperation();
+            }
+        }
+
+
+        public static object add(object a, object b)
+        {
+            if (a is double ad && b is double bd)
+            {
+                return ad + bd;
+            }
+
+            if (a is int ai && b is int bi)
+            {
+                return HandleOverflow(() => checked(ai + bi), () => ai + bi);
+            }
+
+            if ((a is long || a is int) && (b is long || b is int))
+            {
+                var al = Convert.ToInt64(a);
+                var bl = Convert.ToInt64(b);
+                return HandleOverflow(() => checked(al + bl), () => al + bl);
+            }
+
+            if ((a is long || a is double || a is int) && (b is long || b is double || b is int))
+            {
+                return Convert.ToDouble(a) + Convert.ToDouble(b);
```

Comment:
At this point, I believe, either a is double or b is double. Do we really need to convert the other number to double before we add? Won't it work even without any conversion?